File: 0871 - Minimum Number of Refueling Stops/solution.py
import heapq
class Solution:
    def minRefuelStops(self, target: int, fuel: int, A: List[List[int]]) -> int:
        q = []
        A.append([target, 0])
        for p, f in A:
            while q and fuel < p:
                fuel += -heapq.heappop(q)
            if fuel < p:
                return -1
            heapq.heappush(q, -f)
        return len(A)-len(q)

File: 0871 - Minimum Number of Refueling Stops/solution.c
#define max(a,b) (((a)>(b))?(a):(b))
int minRefuelStops(int target, int fuel, int** A, int ASz, int* AColSz){
    long long dp[501] = {fuel};
    for(int i = 0; i < ASz; i++)
        for(int j = i; j>=0 && dp[j]>=A[i][0]; j--)
            dp[j+1] = max(dp[j+1], dp[j]+A[i][1]);
    for(int i = 0; i <= ASz; i++)
        if(dp[i] >= target) return i;
    return -1;
}